Add ATLAS Open Data v3 reader (atlasopenV3); wire into CLA build & CLI

Seloneeko opened the pull request:
**Summary**
Adds support for ATLAS Open Data v3 (Run-3 style) ntuples. New reader: atlasopenV3.
CLI now accepts:

`CLA <file.root> ATLASODV3 -i <analysis.adl>`


**Changes**

`CLA/atlasopenV3.C `— event loop; builds MUO/ELE/TAU/JET/MET collections.

`analysis_core/atlasopenV3.h` — branch setup (RVec pointers, scalars).

`CLA/CLA.Q` — include + input switch for `ATLASODV3`.

`CLA/Makefile` — links the new source into `CLA.exe`.

**Usage**

Tree name expected: `analysis`.

Branches are `ROOT::VecOps::RVec<...>`.

Keys follow existing readers: `"MUO"`, `"ELE"`, `"TAU"`, `"JET"`, `"FJET"` (placeholder), `"PHO"` (placeholder), `"MET"`, `"Truth"`, `"Track"`, `"Combo"`, `"Constits"`.

**Notes**

Units: quick probe of the first events to detect MeV vs GeV; convert to GeV if needed.

Weights: multiply `mcWeight` with available SFs (`ScaleFactor_PILEUP`, `MUON`, `LepTRIGGER`, `JVT`, `BTAG`); missing branches default to 1.

Jets: stash `jet_btag_quantile` in `setFlavor(q)` and set `set_isbtagged_77(q >= 0.77) `as a simple default.

Triggers: wire `trigE` and `trigM` into `evt_data`.

Keeps any framework `TTreeReader` in sync.

**Quick test**
ADL:

```
region SmokeV3
  select size(MUO) >= 1
  histo mu_pt, "mu_pt [GeV]", 50, 0, 200, pt(MUO[0])

```

**Run:**

`CLA ODEO_FEB2025_v0_2muons_mc_*.root ATLASODV3 -i adl_files/smoke_v3.adl`


Then in ROOT:

```
TFile f("histoOut-smoke_v3.root");
((TH1*)f.Get("SmokeV3/mu_pt"))->Draw();
```


**Follow-ups (separate PRs welcome)**

Fill photons / large-R jets / truth once v3 branches are finalized.

Replace the simple b-tag proxy with official WPs when available.

Add a small README snippet + a fuller v3 example ADL.

ahmetcansansar commented:
Hello @Seloneeko, thanks for the nice work. I have a few comments/suggestions for your implementation:

- We would prefer if you use the skeleton from ROOT's `MakeClass("<your tree>")` method in `atlasopenV3.h` . This will properly set all branch names and adresses inside the ntuple, allowing us to fill additional vectors later (photons, large-R jets and truth objects in this case) without modifying the header file. This is valid even for RVec branches, you can check `VLLf.h` and `VLLf.C` as an example.

-  Your style in `atlsaopenV3.C` is a bit different than what we did with the other input types. I don't think this is wrong, but somehow with this event loop setup, CutLang's tree reader function does not work properly for vectors. Maybe it would be good if you stick with the usual style for now, while we understand what was wrong with `atlasopenV3` and tree reader. You can check object definitions in [this](https://github.com/ADL4HEP/ADL4LHCOpenData/blob/main/ATLAS-OTRC-2019-01/ZBosonAnalysis.adl) file to see how we are reading the object attributes straight from the tree. It is very helpful since it allows us to use branches in a tree without setting keywords for every different attribute we see.

-  There is a small mistake in `set_isbtagged_77` attribute. You are using `jet_btag_quantile` when selecting the b-tagged jets at 77% WP. In fact type  of `jet_btag_quantile` branch is integer, and takes values from 1 to 5 (100%, 85%, 77%, 70%, 60%). You should update this as `jtmp.set_isbtagged_77(q >= 3);`

-  We should not set the event weights implicitly in our .C files like you did in Line 283 of `atlasopenV3.C` . The reason behind that is we want to have as much analysis-related information as possible in the .adl files for a clear "analysis description (For your information: That expression looks ok in general, but you also need to divide them by "sum_of_weights" of the event). You can check [this](https://github.com/ADL4HEP/ADL4LHCOpenData/blob/7270df763e8ff0a73a46cbe73c8589b9bab30d71/ATLAS-OTRC-2019-01/ZBosonAnalysis.adl#L37) example to see how to apply weights in .adl file thanks to the tree reader functionality.

Feel free to let me know if you have any questions.

ahmetcansansar commented:
I still don't see the attachments unfortunately. Maybe it is because I am getting your emails as notifications. Could you try sending them directly to my email address (ahmetcansansar1@gmail.com) ?

Seloneeko commented:
I just resent them to your  ***@***.***) email address

On Tue, Sep 9, 2025 at 7:08 PM ahmetcansansar ***@***.***>
wrote:

> *ahmetcansansar* left a comment (unelg/CutLang#169)
> <https://github.com/unelg/CutLang/pull/169#issuecomment-3271376821>
>
> I still don't see the attachments unfortunately. Maybe it is because I am
> getting your emails as notifications. Could you try sending them directly
> to my email address ***@***.***) ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unelg/CutLang/pull/169#issuecomment-3271376821>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BHSR7AIHKVB22GX7RKOEOYL3R33ODAVCNFSM6AAAAACFYAWVXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENZRGM3TMOBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ahmetcansansar commented:
Hi @Seloneeko,

For you compilation issue, maybe you can try compiling inside CLA directory using just "make" instead of using multicore. We are experiencing this issue during clean setups. Please let me know if you can't solve it.

For your implementation, I still have a few comments: 

- I have been able to compile your code, but it was crashing just before the event loop. I've done some debugging on this and looks like using "inline" method causing this crash. It was fixed after I change some lines.

- There are some important details missing in your .C code, most important one: Controller needs that second argument `AnalysisController controller(&aselect, syst_names);` otherwise it stucks at infinite loop when creating output histoOut files. 

- Also please add the following items in your .C code:

1. We have useful print statements in our .C files, like monitoring processed events during execution. These are missing in your version.
2. Add "startevent" "lastevent" variables. This allows us to use `-e` argument to set the number of events we want to run. You can check atlasopenR2 for those.
3. ttr_map does not work properly unless we set particle indices with `setParticleIndx(i)` for each analysis object. 

I've modified your files to make them work. You can continue your work on these. Here is a link to the files: https://cernbox.cern.ch/s/LMvxgJqILhvGG5q

Seloneeko commented:
Hello,
I could compile when I used "make" inside CLA thank you so much. I am not
sure but the issues may be because I am using wsl and there seemed to be
some linking issue when I was trying to compile with CLA_compile yesterday.
I also believe I have fixed the issues you mentioned I will be pushing to
github shortly.

On Wed, Sep 10, 2025 at 2:33 AM ahmetcansansar ***@***.***>
wrote:

> *ahmetcansansar* left a comment (unelg/CutLang#169)
> <https://github.com/unelg/CutLang/pull/169#issuecomment-3272607884>
>
> Hi @Seloneeko <https://github.com/Seloneeko>,
>
> For you compilation issue, maybe you can try compiling inside CLA
> directory using just "make" instead of using multicore. We are experiencing
> this issue during clean setups. Please let me know if you can't solve it.
>
> For your implementation, I still have a few comments:
>
>    -
>
>    I have been able to compile your code, but it was crashing just before
>    the event loop. I've done some debugging on this and looks like using
>    "inline" method causing this crash. It was fixed after I change some lines.
>    -
>
>    There are some important details missing in your .C code, most
>    important one: Controller needs that second argument AnalysisController
>    controller(&aselect, syst_names); otherwise it stucks at infinite loop
>    when creating output histoOut files.
>    -
>
>    Also please add the following items in your .C code:
>
>
>    1. We have useful print statements in our .C files, like monitoring
>    processed events during execution. These are missing in your version.
>    2. Add "startevent" "lastevent" variables. This allows us to use -e
>    argument to set the number of events we want to run. You can check
>    atlasopenR2 for those.
>    3. ttr_map does not work properly unless we set particle indices with
>    setParticleIndx(i) for each analysis object.
>
> I've modified your files to make them work. You can continue your work on
> these. Here is a link to the files:
> https://cernbox.cern.ch/s/LMvxgJqILhvGG5q
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unelg/CutLang/pull/169#issuecomment-3272607884>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BHSR7APUOBVZNPK7KX7WA233R5PTNAVCNFSM6AAAAACFYAWVXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENZSGYYDOOBYGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
